### PR TITLE
🐛🎨♻️Bugfix-refactor: Docker pull progress

### PR DESCRIPTION
**Pull request by sanderegg**
## What do these changes do?
This PR refactors completely the code used by the dynamic-sidecar to pull docker images.

- The code is now located in service-library so that it can be later also used by the computational sidecar
- prior to pulling it calls some special API endpoint on the Docker Registry / Docker Hub to get the image layer information which contain the sizes --> the progress will not anymore go back and forth
- fixes the "Extracting" part that was not shown anymore due to reporting >1.0 progresses

image

Bonuses:
- fix the docker swarm default pool addresses to not overlap with wirguard addresses

## Dev Checklist

- [x] No ENV changes or I properly updated ENV ([read the instruction](https://git.speag.com/oSparc/osparc-ops-deployment-configuration/-/blob/configs/README.md?ref_type=heads#how-to-update-env-variables))

### CODE REVIEW

**Review by GitHK — APPROVED** [review]
thanks for fixing this

**Review by matusdrobuliak66 — APPROVED** [review]

File: packages/service-library/src/servicelib/docker_utils.py [hunk 1]
```diff
@@ -12,3 +28,188 @@ def to_datetime(docker_timestamp: str) -> datetime:
...
+    # NOTE: docker pulls an image layer by layer
+    # NOTE: each layer is first downloaded, then extracted. Extraction usually takes about 2/3 of the time
+    # NOTE: 1 subprogress per layer, then subprogress of size 2 with weights?
+    # NOTE: or, we compute the total size X2, then upgrade that based on the status? maybe simpler
+
+    @dataclass
+    class PulledStatus:
+        size: int
+        downloaded: int = 0
+        extracted: int = 0
```
[the remark this answers is not in the document]
Comment (sanderegg) [reply packages/service-library/src/servicelib/docker_utils.py:144]: done

File: packages/service-library/src/servicelib/docker_utils.py [hunk 2]
```diff
@@ -12,3 +28,188 @@ def to_datetime(docker_timestamp: str) -> datetime:
...
+                    assert parsed_progress.progress_detail.current  # nosec
+                    layer_id_to_size[
+                        parsed_progress.id
+                    ].extracted = parsed_progress.progress_detail.current
+                case "pull complete":
+                    assert parsed_progress.id  # nosec
+                    layer_id_to_size[parsed_progress.id].extracted = layer_id_to_size[
+                        parsed_progress.id
+                    ].size
+                case _:
```
Comment (bisgaard-itis) [inline packages/service-library/src/servicelib/docker_utils.py:199]: So do I understand correctly that this fixes the progress bar for the pulling of the docker images (meaning the data which is part of the service, as opposed to the user data)?

File: packages/service-library/src/servicelib/fastapi/docker_utils.py [hunk 3]
```diff
@@ -0,0 +1,84 @@
...
+from ..docker_utils import (
+    DOCKER_HUB_HOST,
+    DockerImageManifestsV2,
+    DockerImageMultiArchManifestsV2,
+    get_image_complete_url,
+    get_image_name_and_tag,
+)
+
+
+async def retrieve_image_layer_information(
```
Comment (bisgaard-itis) [inline packages/service-library/src/servicelib/fastapi/docker_utils.py:17]: nice!

File: packages/service-library/src/servicelib/docker_utils.py (hunk 2 above)
Comment (bisgaard-itis) [reply packages/service-library/src/servicelib/docker_utils.py:199]: With these changes does the progressbar progress smoothly? Because if that's the case we could give a ETA. I think that's one of the points the POs were pushing for in the last review meetings.

**Review by bisgaard-itis — APPROVED** [review]
Very cool! Thanks. I have a few questions.

File: packages/service-library/src/servicelib/docker_utils.py (hunk 2 above)
Comment (bisgaard-itis) [reply packages/service-library/src/servicelib/docker_utils.py:200]: Cool, thanks for the clarification. Yes, I think it would be great if we did some statistics so that we could maybe create a "smooth" progress bar together with a time estimate at some point

File: Makefile [hunk 4]
```diff
@@ -396,8 +396,8 @@ leave: ## Forces to stop all services, networks, etc by the node leaving the swa
 
 .PHONY: .init-swarm
 .init-swarm:
-	# Ensures swarm is initialized (careful we use a default pool of 172.16.0.1/16. Ensure you do not use private IPs in that range!)
-	$(if $(SWARM_HOSTS),,docker swarm init --advertise-addr=$(get_my_ip) --default-addr-pool 172.16.0.1/16)
+	# Ensures swarm is initialized (careful we use a default pool of 10.20.0.0/16. Ensure you do not use private IPs in that range!)
```
Comment (pcrespov) [inline Makefile:399]: I would like to learn more about this offline. thx

File: packages/aws-library/requirements/dev.txt [hunk 5]
```diff
@@ -12,7 +12,7 @@
 --requirement _tools.txt
 
 # installs this repo's packages
-pytest-simcore @ ../pytest-simcore
```
Comment (pcrespov) [inline packages/aws-library/requirements/dev.txt:15]: good catch. Did you check that we did not do elsewhere? :-)

File: packages/pytest-simcore/src/pytest_simcore/docker_registry.py [hunk 6]
```diff
@@ -285,3 +293,17 @@ def dy_static_file_server_dynamic_sidecar_compose_spec_service(
         docker_registry,
         node_meta_schema,
     )
+
+
+@pytest.fixture
+def remove_images_from_host() -> Callable[[list[str]], Awaitable[None]]:
+    async def _cleaner(images: list[str]) -> None:
+        print(f"--> Removing {images}")
+        async with aiodocker.Docker() as client:
```
Comment (pcrespov) [inline packages/pytest-simcore/src/pytest_simcore/docker_registry.py:302]: Q: AFAIK we are using both `docker` and `aiodocker`. Should we priorize using the latter?

File: packages/pytest-simcore/src/pytest_simcore/docker_registry.py [hunk 7]
```diff
@@ -285,3 +293,17 @@ def dy_static_file_server_dynamic_sidecar_compose_spec_service(
         docker_registry,
         node_meta_schema,
     )
+
+
+@pytest.fixture
+def remove_images_from_host() -> Callable[[list[str]], Awaitable[None]]:
+    async def _cleaner(images: list[str]) -> None:
+        print(f"--> Removing {images}")
```
Comment (pcrespov) [inline packages/pytest-simcore/src/pytest_simcore/docker_registry.py:301]: MINOR: Can use the new `log_context` here instead of the `print`.

File: tests/public-api/requirements/_base.txt [hunk 8]
```diff
@@ -1,5 +1,5 @@
 # This file was autogenerated by uv via the following command:
-#    uv pip compile --output-file requirements/_base.txt requirements/_base.in
+#    uv pip compile --upgrade-package aiodocker --output-file requirements/_base.txt requirements/_base.in
```
Comment (pcrespov) [inline tests/public-api/requirements/_base.txt:2]: undo this change. It tracked that you upgraded a single package. It is the correct way to go but we should not upgrade base.txt for this. 
I wonder if uv has some options to avoid these changes otherwise it is going to be a bit more time consuming upgrading packages

File: services/web/server/docker/boot.sh [hunk 9]
```diff
@@ -63,5 +64,6 @@ else
     --workers="${WEBSERVER_GUNICORN_WORKERS:-1}" \
     --name="webserver_$(hostname)_$(date +'%Y-%m-%d_%T')_$$" \
     --access-logfile='-' \
-    --access-logformat='%a %t "%r" %s %b [%Dus] "%{Referer}i" "%{User-Agent}i"'
+    --access-logformat='%a %t "%r" %s %b [%Dus] "%{Referer}i" "%{User-Agent}i"' \
+    --worker-tmp-dir=/dev/shm
```
Comment (pcrespov) [inline services/web/server/docker/boot.sh:68]: can you elaborate a bit on why using `/dev/shm` ant not e.g. `/tmp` for the boot script. Or simply drop a comment here thx :-)

File: services/dynamic-sidecar/tests/unit/test_api_containers.py [hunk 10]
```diff
@@ -692,7 +692,9 @@ async def test_attach_detach_container_to_network(
                 container_inspect = await container.show()
                 networks = container_inspect["NetworkSettings"]["Networks"]
                 assert network_id in networks
-                assert set(networks[network_id]["Aliases"]) == set(network_aliases)
+                assert set(network_aliases).issubset(
```
Comment (pcrespov) [inline services/dynamic-sidecar/tests/unit/test_api_containers.py:695]: Haven't I seen this one in another PR :-)?

File: packages/service-library/tests/aiohttp/test_docker_utils.py [hunk 11]
```diff
@@ -0,0 +1,120 @@
...
+    registry_settings: RegistrySettings,
+    osparc_service: dict[str, Any],
+    service_repo: str,
+    service_tag: str,
+):
+    # clean first
+    image_name = f"{service_repo}:{service_tag}"
+    if "sha256" in service_tag:
+        image_name = f"{service_repo}@{service_tag}"
+    await remove_images_from_host([image_name])
```
Comment (pcrespov) [inline packages/service-library/tests/aiohttp/test_docker_utils.py:38]: why not testing that they are actually removed? e.g. checking that they are not listed

File: packages/service-library/tests/aiohttp/test_docker_utils.py [hunk 12]
```diff
@@ -0,0 +1,120 @@
...
+            image, registry_settings, main_progress_bar, fake_log_cb, layer_information
+        )
+        fake_log_cb.assert_called()
+        assert main_progress_bar._current_steps == 1  # noqa: SLF001
+    assert fake_progress_report_cb.call_args_list[0] == call(0.0)
+    fake_progress_report_cb.assert_called_with(1.0)
+
+    # check there were no warnings
+    for record in caplog.records:
+        assert record.levelname != "WARNING", record.message
```
Comment (pcrespov) [inline packages/service-library/tests/aiohttp/test_docker_utils.py:104]: what about ERRORS?

File: packages/service-library/tests/fastapi/test_docker_utils.py [hunk 13]
```diff
@@ -0,0 +1,120 @@
...
+)
+async def test_retrieve_image_layer_information(
+    remove_images_from_host: Callable[[list[str]], Awaitable[None]],
+    registry_settings: RegistrySettings,
+    osparc_service: dict[str, Any],
+    service_repo: str,
+    service_tag: str,
+):
+    # clean first
+    image_name = f"{service_repo}:{service_tag}"
```
Comment (pcrespov) [inline packages/service-library/tests/fastapi/test_docker_utils.py:35]: THOUGHT: you have used this resolution of the image_name in several places. If you are using it in the code perhaps it would be a good idea to have an util that composes image names from parts? Similar to types like HttpUrl from pydantic

File: packages/service-library/tests/fastapi/test_docker_utils.py
Comment (pcrespov) [inline packages/service-library/tests/fastapi/test_docker_utils.py:1]: I did not know one could use the same filename in tests (even in different folders)! I guess you have no problems because we run tests for `aiohttp` and `fastapi` folders separately. Otherwise I bet there might be issues.

**Review by pcrespov — APPROVED** [review]
Great job! Left some suggestions and comments!

File: packages/aws-library/requirements/dev.txt (hunk 5 above)
Comment (sanderegg) [reply packages/aws-library/requirements/dev.txt:15]: no "we" did not ;)

File: tests/public-api/requirements/_base.txt (hunk 8 above)
Comment (sanderegg) [reply tests/public-api/requirements/_base.txt:2]: ok.

File: services/web/server/docker/boot.sh (hunk 9 above)
Comment (sanderegg) [reply services/web/server/docker/boot.sh:68]: https://pythonspeed.com/articles/gunicorn-in-docker/

File: packages/service-library/tests/fastapi/test_docker_utils.py
Comment (sanderegg) [reply packages/service-library/tests/fastapi/test_docker_utils.py:1]: yeah actually it is a bit sad. it would nicer to have a common interface. implemented by httpx or aiohttp, cause that would prevent duplicating code.
same goes with long running tasks actually

File: packages/service-library/tests/aiohttp/test_docker_utils.py (hunk 12 above)
Comment (sanderegg) [reply packages/service-library/tests/aiohttp/test_docker_utils.py:104]: added notes

File: packages/pytest-simcore/src/pytest_simcore/docker_registry.py (hunk 7 above)
Comment (sanderegg) [reply packages/pytest-simcore/src/pytest_simcore/docker_registry.py:301]: actually I am wondering why the regular one is not also upgraded with the same options

File: packages/pytest-simcore/src/pytest_simcore/docker_registry.py (hunk 6 above)
Comment (sanderegg) [reply packages/pytest-simcore/src/pytest_simcore/docker_registry.py:302]: hmm... actually did not know when I started I would have also to move the log_context away from playwright and then fix playwright. hem. but ok

File: packages/service-library/tests/fastapi/test_docker_utils.py (hunk 13 above)
Comment (sanderegg) [reply packages/service-library/tests/fastapi/test_docker_utils.py:35]: not sure I understand this comment

File: packages/service-library/tests/aiohttp/test_docker_utils.py (hunk 11 above)
Comment (sanderegg) [reply packages/service-library/tests/aiohttp/test_docker_utils.py:38]: done

File: packages/pytest-simcore/src/pytest_simcore/docker_registry.py (hunk 7 above)
Comment (sanderegg) [reply packages/pytest-simcore/src/pytest_simcore/docker_registry.py:301]: now I know...
